### ISSUE TITLE
Email Notification When Stock Level Drops Below Minimum?

### ISSUE THREAD

**Issue by skydiablo** [inventree/InvenTree#8452]
### Please verify that this bug has NOT been raised before.

- [x] I checked and didn't find a similar issue

### Describe the bug*

I'm not receiving email notifications when a part's stock level drops below the set minimum, even though email functionality is working, and notifications are enabled in the user settings.

Could this issue (https://github.com/inventree/InvenTree/issues/7866) be addressing the same question?

### Steps to Reproduce

1. Set a minimum stock level for a part.
1. Ensure email notifications are enabled in the user settings.
1. Allow the stock level of that part to drop below the minimum.

### Expected behaviour

An email notification should be sent when the stock level falls below the minimum threshold.

### Deployment Method

- [ ] Docker
- [ ] Package
- [x] Bare metal
- [ ] Other - added info in Steps to Reproduce

### Version Information

# Version Information:
InvenTree-Version: 0.16.8
Django Version: 4.2.15

Database: sqlite3
Debug-Mode: False
Deployed using Docker: False
Platform: Linux-6.8.12-2-pve-x86_64-with-glibc2.31
Installer: None

Active plugins: [{'name': 'InvenTreeBarcode', 'slug': 'inventreebarcode', 'version': '2.1.0'}, {'name': 'InvenTreeCoreNotificationsPlugin', 'slug': 'inventreecorenotificationsplugin', 'version': '1.0.0'}, {'name': 'InvenTreeCurrencyExchange', 'slug': 'inventreecurrencyexchange', 'version': '1.0.0'}, {'name': 'InvenTreeLabel', 'slug': 'inventreelabel', 'version': '1.1.0'}, {'name': 'InvenTreeLabelMachine', 'slug': 'inventreelabelmachine', 'version': '1.0.0'}, {'name': 'InvenTreeLabelSheet', 'slug': 'inventreelabelsheet', 'version': '1.0.0'}, {'name': 'DigiKeyPlugin', 'slug': 'digikeyplugin', 'version': '1.0.0'}, {'name': 'LCSCPlugin', 'slug': 'lcscplugin', 'version': '1.0.0'}, {'name': 'MouserPlugin', 'slug': 'mouserplugin', 'version': '1.0.0'}, {'name': 'TMEPlugin', 'slug': 'tmeplugin', 'version': '1.0.0'}, {'name': 'IPNGenerator', 'slug': 'ipngen', 'version': '0.1'}]

### Please verify if you can reproduce this bug on the demo site.

- [ ] I can reproduce this bug on the demo site.

**Comment by skydiablo** [comment]
Alright, I had a notification set on the category, but no notification was triggered on the web interface, nor did I receive an email. When I set the notification directly on the "part," it works as expected. It seems like setting it on the category doesn’t work?

Workflow question: Regarding variants, I would like to receive a notification whenever an individual variant falls below the minimum stock level. If I understand correctly, I now have to activate the notification for each part/variant individually? If I set a minimum stock level on the reference part, a notification is only triggered when the total of all variants falls below the minimum, which makes sense, but how can I receive a notification for each variant without activating them all individually? And setting it on the category doesn’t work at all?

**Comment by SchrodingersGat** [comment]
@skydiablo this seems related to these fixes, maybe? 

- https://github.com/inventree/InvenTree/pull/8360
- https://github.com/inventree/InvenTree/pull/8361

> When I set the notification directly on the "part," it works as expected. It seems like setting it on the category doesn’t work?

Could you try to recreate this across a number of different part / category combinations? The notifcation based on subscibred category should be working fine?

**Comment by skydiablo** [comment]
hmmm, i have try to trigger a new notification, no success :( independed how i subscribe to the part. no notification at all. is there any dependecies how long a state have to be, between add and remove the stock items?

**Comment by skydiablo** [comment]
- subscribe to most top part category
- crate a new part in the subscribed category (as template = true) 
- add a variant
- add a stock
- define minimum limit to the variant  (10)
- add stock (10)
- remove stock (5)
--> nothing happen :(

same as above but remove the category subscription and subscribe direct the variant --> nothing happen :(

**Comment by SchrodingersGat** [comment]
You should be able to see if the notification entry has been created, in the admin interface, e.g. http://localhost:8000/admin/common/notificationentry/

(replace "localhost:8000" with your server name)

For example:

![Image](https://github.com/user-attachments/assets/24c4137b-c789-4df4-afbb-39c328c25149)

**Comment by skydiablo** [comment]
it seems that i get one notification per day at all?! if i get an notification for low part stock, i will never get again this day a another notification in any other combination of subscribtions.

**Comment by SchrodingersGat** [comment]
@skydiablo one notification on any "topic"? The rate limit should only apply to ensure that you don't receive multiple notifications per day for an individual part. But, if more than one part goes "low stock" in a day you should receive all notifications. Can you confirm?

**Comment by skydiablo** [comment]
no, i get one notification for the first part goes "low stock", if there any other part goes in "low stock" state at the same day, no notification is send anymore.

**Comment by SchrodingersGat** [comment]
Do you receive multiple notifications in the web interface? Or only a single one there also?

**Comment by skydiablo** [comment]
just a single one at all

**Comment by skydiablo** [comment]
just one per day:
![Image](https://github.com/user-attachments/assets/cd591d93-a2bb-49ad-91f7-b8e3f51bf8df)

**Comment by SchrodingersGat** [comment]
I have just tested this and notifications are correctly generated for different parts, one after the other:

![Image](https://github.com/user-attachments/assets/07327c41-abbf-470e-86ad-c80adc7f6dfe)

![Image](https://github.com/user-attachments/assets/c139502c-63c8-4ef3-8930-4d153a06716f)

![Image](https://github.com/user-attachments/assets/e7a2d392-934c-4619-aae1-2509e38d77e1)

Can you please edit a bunch of parts, such that the "minimum stock" is higher than the actual stock, and then share screenshots of the "notification entry" and "notification message" admin pages (like I have shared above).

**Comment by skydiablo** [comment]
i had read some code and it seems there is maybe a bug. the function to call to check is some part in "low stock" is [notify_low_stock_if_required](https://github.com/inventree/InvenTree/blob/5e762bc7f7c873c53240ededb8f7dba13795dab1/src/backend/InvenTree/part/tasks.py#L54). this will called on some DB events, also like this [after_save_stock_item](https://github.com/inventree/InvenTree/blob/5e762bc7f7c873c53240ededb8f7dba13795dab1/src/backend/InvenTree/stock/models.py#L2419). this is the function that will call if the stock-count will be changed also, right? but only if this a new object in database, like `created==true` ([here](https://github.com/inventree/InvenTree/blob/5e762bc7f7c873c53240ededb8f7dba13795dab1/src/backend/InvenTree/stock/models.py#L2423)).
the event `notify_low_stock_if_required` is also called if the part is saved in [after_save_part](https://github.com/inventree/InvenTree/blob/5e762bc7f7c873c53240ededb8f7dba13795dab1/src/backend/InvenTree/part/models.py#L2551). but here is the trigger called only if `created==false`.
in my opinion, in function [after_save_stock_item](https://github.com/inventree/InvenTree/blob/5e762bc7f7c873c53240ededb8f7dba13795dab1/src/backend/InvenTree/stock/models.py#L2419), the `created` is not to check. and this is the reason why i do not get an event if i change the stock count. if i also change the stock count under the threshold and also change the part itself and save the part, i will get a notification for "low stock".

what do you think @SchrodingersGat ?

**Comment by SchrodingersGat** [comment]
@skydiablo looks like you are right - saving a stock item should trigger this event, whether it is a newly created stock item or editing an existing one.

Would you be able to submit a PR to address this? We should also extend the existing unit testing to ensure that this does not get changed again
